src/main/java/com/google/devtools/build/lib/bazel/bzlmod/GsonTypeAdapterUtil.java

```diff
@@ -444,6 +445,20 @@ public Location read(JsonReader jsonReader) throws IOException {
     }
   }
 
+  private static final TypeAdapter<RepoRecordedInput.File> REPO_RECORDED_INPUT_FILE_TYPE_ADAPTER =
```

Comment:
I have the same vague dissatisfaction with this as with time stamps in lockfiles: this encodes non-hermetic information in lockfiles, which are supposed to be checked in. Although in this case, one could say that that could be avoid by carefully not depending on absolute paths...

Reply:
Now that we have two mechanisms for non-hermetic operations, repository rules and module extensions, we should think about where we want to draw the boundary between them. If module extensions are meant to be optimized for "reproducibility in the face of non-hermeticity", i.e., locking down the result of potentially non-hermetic operations in a way that is portable across machines via the lockfile, then we may want to purposefully limit their capabilities. 

How about adding the watch capability to repo rules only for now? We can always gather some data from user feedback before we extend it to module extensions.

Reply:
good points, and I'm not totally sure what's the best way forward here. It's awkward to add watch only to repo rules because `rctx.read(watch=True)` is in `StarlarkBaseExternalContext` which is shared by both contexts. We could alternatively just filter out any "absolute file" watches when we write the lockfile. WDYT?

Reply:
I'm not very knowledgeable about the philosophical difference between module extensions and repositories, but at least as far as the implementation goes, it'd be pretty easy to add an "access to absolute paths is allowed" flag to `StarlarkBaseExternalContext`, which would be set to "true" for repositories and "false" for module extensions.

Between filtering out absolute file watches and erroring out when they are attempted, I'd rather we do the latter because if we ever decide that this is a useful thing, it's much more easy to turn an error into a non-error than to subtly change the behavior of code that Bazel already accepts.

Reply:
the situation is a bit more nuanced -- I'm adding a boolean parameter `watch` to `rctx.read` which defaults to True. That is, any file read by `rctx.read` will be watched, be it a label-addressed file or an absolute path file.

Now we can't really just make it an error to attempt to watch absolute path files in module extensions, because it _is_ legal to read an absolute path file in module extensions today. And yes, we could make `watch` default to True for repo rules and False for module extensions, but I'm not sure that's better...

Reply:
my newest idea: make `rctx.read(watch)` a tristate: true - always watch (same as today's `rctx.read(...); rctx.watch(...);`); false - never watch; auto - watch if appropriate ("appropriate" = it's a file inside the workspace OR we're in a repo rule context). Defaults to 'auto'. WDYT?

Reply:
WDYT about `watch()` defaulting to false everywhere, at least for the time being? That would mean that there is no behavior change and it would also resolve the module extension / repository conundrum.

re: your tristate approach, I'm neutral; I think it's a bit of DWIM, but only a bit so I could live with it.

Reply:
I wanted to enable watch by default because the only change in behavior is that we refetch a bit more often, which is really an improvement in correctness. It also reduces surprises -- people have been asking why label-addressed files are watched but others aren't.